sync3/caches/global.go

```diff
@@ -76,49 +77,73 @@ func (c *GlobalCache) OnRegistered(_ context.Context, _ int64) error {
 	return nil
 }
 
-// Load the current room metadata for the given room IDs. Races unless you call this in a dispatcher loop.
+// LoadRooms loads the current room metadata for the given room IDs. Races unless you call this in a dispatcher loop.
 // Always returns copies of the room metadata so ownership can be passed to other threads.
-// Keeps the ordering of the room IDs given.
 func (c *GlobalCache) LoadRooms(ctx context.Context, roomIDs ...string) map[string]*internal.RoomMetadata {
 	c.roomIDToMetadataMu.RLock()
 	defer c.roomIDToMetadataMu.RUnlock()
 	result := make(map[string]*internal.RoomMetadata, len(roomIDs))
 	for i := range roomIDs {
 		roomID := roomIDs[i]
-		sr := c.roomIDToMetadata[roomID]
-		if sr == nil {
-			logger.Warn().Str("room", roomID).Msg("GlobalCache.LoadRoom: no metadata for this room, generating stub")
-			c.roomIDToMetadata[roomID] = internal.NewRoomMetadata(roomID)
-			sr = c.roomIDToMetadata[roomID]
-		}
-		srCopy := *sr
-		// copy the heroes or else we may modify the same slice which would be bad :(
-		srCopy.Heroes = make([]internal.Hero, len(sr.Heroes))
-		for i := range sr.Heroes {
-			srCopy.Heroes[i] = sr.Heroes[i]
-		}
-		result[roomID] = &srCopy
+		result[roomID] = c.copyRoom(roomID)
 	}
 	return result
 }
 
-// Load all current joined room metadata for the user given. Returns the absolute database position along
-// with the results. TODO: remove with LoadRoomState?
-func (c *GlobalCache) LoadJoinedRooms(ctx context.Context, userID string) (pos int64, joinedRooms map[string]*internal.RoomMetadata, err error) {
+// LoadRoomsFromMap is like LoadRooms, except it is given a map with room IDs as keys.
+// The values in that map are completely ignored.
+func (c *GlobalCache) LoadRoomsFromMap(ctx context.Context, joinNIDsByRoomID map[string]int64) map[string]*internal.RoomMetadata {
+	c.roomIDToMetadataMu.RLock()
+	defer c.roomIDToMetadataMu.RUnlock()
+	result := make(map[string]*internal.RoomMetadata, len(joinNIDsByRoomID))
+	for roomID, _ := range joinNIDsByRoomID {
+		result[roomID] = c.copyRoom(roomID)
+	}
+	return result
+}
+
+// copyRoom returns a copy of the internal.RoomMetadata stored for this room.
+// This is an internal implementation detail of LoadRooms and LoadRoomsFromMap.
+// If the room is not present in the global cache, returns a stub metadata entry.
+// The caller MUST acquire a read lock on roomIDToMetadataMu before calling this.
+func (c *GlobalCache) copyRoom(roomID string) *internal.RoomMetadata {
+	sr := c.roomIDToMetadata[roomID]
+	if sr == nil {
+		logger.Warn().Str("room", roomID).Msg("GlobalCache.LoadRoom: no metadata for this room, generating stub")
+		c.roomIDToMetadata[roomID] = internal.NewRoomMetadata(roomID)
```

Comment:
Comments on this function suggest a read lock is needed, but here we write.

Reply:
Oh true. An error in https://github.com/matrix-org/sliding-sync/pull/136 then?

Reply:
Thought process:
- could acquire a write lock in the caller, but I'm guessing LoadRooms is called frequently, and don't want to nobble concurrent reads
- maybe we just return a blank stub but don't insert it into the map?